[Feature]: Add model docs (#442)
open-mmlab/mmselfsup: docs/en/advanced_guides/models.md

# Models

- [Models](#models)
  - [Overview of modules in MMSelfSup](#overview-of-modules-in-mmselfsup)
  - [Construct algorithms from sub-modules](#construct-algorithms-from-sub-modules)
  - [Overview these abstract functions in base model](#overview-these-abstract-functions-in-base-model)

Model can be seen as a feature extractor or loss generator for each algorithm. In MMSelfSup, it mainly
contains the following fix parts,

- algorithms, containing the full modules of a model and all sub-modules will be
  constructed in algorithms.
- backbones, containing the backbones for each algorithm, e.g. ViT for MAE, and Swim Transformer
  for SimMIM.
- necks, some specifial modules, such as decoder, appended directly to the output of the backbone.
- heads, some specifial modules, such as mlp layers, appended to the output of the backbone or neck.

> - memories, some memory banks or queues in some algorithm, e.g. MoCov1/v2.

Reviewer: ```suggestion
- memories, some memory banks or queues in some algorithms, e.g. MoCov1/v2.
```

| [`RotationPred`](TODO) |        [`ResNet`](TODO)         |             N/A              |          [`ClsHead`](TODO)           |     [`CrossEntropyLoss`](TODO)     |          N/A           |
|    [`SimCLR`](TODO)    |        [`ResNet`](TODO)         |   [`NonLinearNeck`](TODO)    |      [`ContrastiveHead`](TODO)       |     [`CrossEntropyLoss`](TODO)     |          N/A           |
|    [`SimMIM`](TODO)    | [`SimMIMSwinTransformer`](TODO) |     [`SimMIMNeck`](TODO)     |         [`SimMIMHead`](TODO)         | [`SimMIMReconstructionLoss`](TODO) |          N/A           |
|   [`SimSiam`](TODO)    |        [`ResNet`](TODO)         |   [`NonLinearNeck`](TODO)    |     [`LatentPredictHead`](TODO)      |   [`CosineSimilarityLoss`](TODO)   |          N/A           |
|     [`SwAV`](TODO)     |        [`ResNet`](TODO)         |      [`SwAVNeck`](TODO)      |          [`SwAVHead`](TODO)          |         [`SwAVLoss`](TODO)         |          N/A           |

## Construct algorithms from sub-modules

> Just as shown in above table, each algorithm is a combination of backbone, neck, head, loss, and memories. You are free to use these existing modules to build your own algorithms. If there are some customized modules, you should follow [add_modules](./add_modules.md)

Reviewer: ```suggestion
Just as shown in above table, each algorithm is a combination of backbone, neck, head, loss and memory. You are free to use these existing modules to build your own algorithms. If the customized modules are required, you are supposed to follow [add_modules](./add_modules.md) to meet your needs.
```

                 pretrained: Optional[str] = None,
                 data_preprocessor: Optional[Union[dict, nn.Module]] = None,
                 init_cfg: Optional[dict] = None):

        if pretrained is not None:
            init_cfg = dict(type='Pretrained', checkpoint=pretrained)

        if data_preprocessor is None:
            data_preprocessor = {}
        # The build process is in MMEngine, so we need to add scope here.
        data_preprocessor.setdefault('type',
                                     'mmselfsup.SelfSupDataPreprocessor')

        super().__init__(
            init_cfg=init_cfg, data_preprocessor=data_preprocessor)

        self.backbone = MODELS.build(backbone)

        if neck is not None:
            self.neck = MODELS.build(neck)

        if head is not None:
            self.head = MODELS.build(head)

```

Just as shown above, you should provide the config to build the backbone, but neck and head are optional. In addition to building
your algorithm, you should overwrite some abstract functions in the base model to get the correct results, which we will discuss in the
following section.

## Overview these abstract functions in base model

> The `forward` function is the entrance to the your results. However, it is different from the default `forward` function in most PyTorch code, which

Reviewer: ```suggestion
The `forward` function is the entrance to the results. However, it is different from the default `forward` function in most PyTorch code, which
```

> only has one mode. You will mess all your logic in the `forward` function, limiting the scability. Just as shown in the table below, `forward` function in MMSelfSup has three modes, i) tensor, ii) loss and iii) predict.

Reviewer: ```suggestion
only has one mode. You will mess all your logic in the `forward` function, limiting the scability. Just as shown in the code below, `forward` function in MMSelfSup has three modes, i) tensor, ii) loss and iii) predict.
```
